File: sdks/python/starfish/heartbeat.py
from __future__ import annotations

import asyncio
import time
from typing import TYPE_CHECKING

from .id import next_id
from .types import StarfishFrame, StarfishHeader

if TYPE_CHECKING:
    from .connection import Connection


class Heartbeat:
    def __init__(self, connection: Connection) -> None:
        self._connection = connection
        self._task: asyncio.Task[None] | None = None

    def start(self) -> None:
        self.stop()
        self._task = asyncio.create_task(self._run())

    def stop(self) -> None:
        if self._task:
            self._task.cancel()
            self._task = None

    async def _run(self) -> None:
        interval = self._connection.heartbeat_interval / 1000
        try:
            while True:
                await asyncio.sleep(interval)
                try:
                    ping = StarfishFrame(
                        header=StarfishHeader(
                            id=next_id("ping"),
                            resource="heartbeat",
                            method="ping",
                            kind="request",
                            ts=int(time.time() * 1000),
                        ),
                    )
                    await self._connection.send(ping)
                except Exception:
                    pass  # connection closed; heartbeat restarts on reconnect
        except asyncio.CancelledError:
            return

File: sdks/golang/starfish/heartbeat.go
package starfish

import (
	"context"
	"time"
)

// heartbeat sends application-level pings at the server-specified interval.
type heartbeat struct {
	conn     *connection
	idg      *IDGenerator
	interval time.Duration
	cancel   context.CancelFunc
}

func newHeartbeat(conn *connection, idg *IDGenerator) *heartbeat {
	return &heartbeat{
		conn: conn,
		idg:  idg,
	}
}

// start begins sending pings at the given interval.
func (h *heartbeat) start(ctx context.Context, intervalMs int) {
	h.stop()

	h.interval = time.Duration(intervalMs) * time.Millisecond
	ctx, cancel := context.WithCancel(ctx)
	h.cancel = cancel

	go h.loop(ctx)
}

func (h *heartbeat) loop(ctx context.Context) {
	ticker := time.NewTicker(h.interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			ts := time.Now().UnixMilli()
			frame := &Frame{
				Header: Header{
					V:        1,
					ID:       h.idg.Next("ping"),
					Resource: "heartbeat",
					Method:   "ping",
					Kind:     "request",
					Ts:       &ts,
				},
			}
			_ = h.conn.send(ctx, frame) // best-effort
		}
	}
}

// stop cancels the heartbeat loop.
func (h *heartbeat) stop() {
	if h.cancel != nil {
		h.cancel()
		h.cancel = nil
	}
}
